File: Donut.cpp
//
// Created by Romanov on 13.11.2021.
//

#include "Donut.h"

int main() {
	float A = 0, B = 0;
	float i, j;
	int k;
	float z[1760];
	char b[1760];
	std::cout << "\x1b[2J";
	while (true) {
		memset(b, 32, 1760);
		memset(z, 0, 7040);
		for (j = 0; j < 6.28; j += 0.07) {
			for (i = 0; i < 6.28; i += 0.02) {
				float c = std::sin(i);
				float d = std::cos(j);
				float e = std::sin(A);
				float f = std::sin(j);
				float g = std::cos(A);
				float h = d + 2;
				float D = 1 / (c * h * e + f * g + 5);
				float l = std::cos(i);
				float m = std::cos(B);
				float n = std::sin(B);
				float t = c * h * g - f * e;
				int x = 40 + 30 * D * (l * h * m - t * n);
				int y = 12 + 15 * D * (l * h * n + t * m);
				int o = x + 80 * y;
				int N = 8 * ((f * e - c * d * g) * m - c * d * e - f * g - l * d * n);
				if (22 > y && y > 0 && x > 0 && 80 > x && D > z[o]) {
					z[o] = D;
					b[o] = ".,-~:;=!*#$@"[N > 0 ? N : 0];
				}
			}
		}
		std::cout << "\x1b[H";
		for (k = 0; k < 1761; k++) {
			putchar(k % 80 ? b[k] : 10);
			A += 0.00004;
			B += 0.00002;
		}
		gotoxy(0, 0);
	}
	return 0;
}

File: Donut.java
package me.tylermoser;

import java.io.IOException;
import java.time.Instant;
import java.util.Arrays;

public class Donut {

    private static final int CANVAS_WIDTH = 640;
    private static final int CANVAS_HEIGHT = 640;
    private static final int CANVAS_LENGTH = CANVAS_WIDTH * CANVAS_HEIGHT;
    private static final float TWO_PI = 6.28318f;
    private static final int RUNTIME_SECS = 60;
    private static final boolean DISPLAY = true;

    public static void main(String[] args) throws InterruptedException, IOException {
        long iter = 0;
        long start = Instant.now().getEpochSecond();
        char[] asciiPixels = {'.', ',', '-', '~', ':', ';', '=', '!', '*', '#', '$', '@'};
        DonutWindow window = DISPLAY ? new DonutWindow(CANVAS_WIDTH, CANVAS_HEIGHT) : null;

        float A = 0, B = 0;
        float phi, theta;
        char[] frameBuffer = new char[CANVAS_LENGTH];
        int[] pixelBuffer = new int[CANVAS_LENGTH];
        float[] zBuffer = new float[CANVAS_LENGTH];
        while (true) {
            A += 0.07; // Increment torus rotation
            B += 0.03; // Increment torus rotation
            Arrays.fill(frameBuffer, ' ');
            Arrays.fill(pixelBuffer, (byte) 0xFF);
            Arrays.fill(zBuffer, 0);
            for (theta = 0; theta < TWO_PI; theta += 0.003) { // Originally 0.07
                for (phi = 0; phi < TWO_PI; phi += 0.001) { // Originally 0.02
                    float sTheta = (float) Math.sin(theta), cTheta = (float) Math.cos(theta);
                    float sPhi = (float) Math.sin(phi), cPhi = (float) Math.cos(phi);
                    float sA = (float) Math.sin(A), cA = (float) Math.cos(A);
                    float sB = (float) Math.sin(B), cB = (float) Math.cos(B);

                    float h = cTheta + 2;
                    float t = sPhi * h * cA - sTheta * sA;
                    float oneOverZ = 1 / (sPhi * h * sA + sTheta * cA + 5);
                    int x2d = (int) (320 + 400 * oneOverZ * (cPhi * h * cB - t * sB));
                    int y2d = (int) (320 + 400 * oneOverZ * (cPhi * h * sB + t * cB));
                    int bufferIndex = x2d + CANVAS_WIDTH * y2d;
                    float illumination = (sTheta * sA - sPhi * cTheta * cA) * cB - sPhi * cTheta * sA - sTheta * cA - cPhi * cTheta * sB;
                    int illumination12 = (int) (8 * illumination); // multiplier = floor( # of levels / sqrt(2) )
                    byte illumination255 = (byte) (90 * illumination);
                    if (CANVAS_HEIGHT > y2d && y2d > 0 && x2d > 0 && CANVAS_WIDTH > x2d && oneOverZ > zBuffer[bufferIndex]) {
                        zBuffer[bufferIndex] = oneOverZ;
                        frameBuffer[bufferIndex] = asciiPixels[Math.max(illumination12, 0)];
                        pixelBuffer[bufferIndex] = 128 + illumination255;
                    }
                }
            }

            if (DISPLAY) {
                //printAsciiDonut(frameBuffer);
                window.updateDonut(pixelBuffer);
            }

            iter++;
            if (Instant.now().getEpochSecond() - start > RUNTIME_SECS) {
                System.out.printf("%f\n", ((double)iter / RUNTIME_SECS));
                return;
            }
        }
    }

    private static void printAsciiDonut(char[] frameBuffer) throws IOException, InterruptedException {
        if (System.getProperty("os.name").contains("Windows")) {
            new ProcessBuilder("cmd", "/c", "cls").inheritIO().start().waitFor();
        } else {
            System.out.print("\033[H\033[2J");
        }
        for (int i = 0; i < CANVAS_LENGTH + 1; i++) {
            System.out.print(i % CANVAS_WIDTH > 0 ? frameBuffer[i] : '\n');
        }
        System.out.flush();
    }

}
